src/components/i18n/FormattedCompMessage.js

```diff
@@ -0,0 +1,205 @@
+import {
+    CATEGORY_ZERO,
+    CATEGORY_ONE,
+    CATEGORY_TWO,
+    CATEGORY_FEW,
+    CATEGORY_MANY,
+    CATEGORY_OTHER,
+} from '../../constants';
+import Composition from './Composition';
+
+type Props = {
+    /**
+     * The text to translate. This may be a string or JSX. The defaultMessage prop may be
+     * given or the component may have children, but not both.
+     */
+    children?: React.Node,
+
+    /**
+     * Specify the pivot count to choose which plural form to use.
+     * When specified, this FormattedCompMessage component will choose one of the
+     * Plural elements in its children according to the value of this count
+     * and the linguistic rules of the locale which determine which numbers
+     * belong to which plural class.
+     */
+    count?: number,
+
+    /**
+     * The text to translate. This may be a string or JSX. This prop may be
+     * given or the component may have children, but not both.
+     */
+    defaultMessage?: React.ElementType | string,
+
+    /**
+     * A description to send to the translators to explain the context of
+     * this string.
+     */
+    description: string,
+
+    /** The unique id of this string. */
+    id: string,
+
+    /**
+     * The intl provider. This is injected into this component
+     * via the injectIntl function from react-intl.
+     */
+    intl: Object,
+
+    /**
+     * Specify the name of the HTML tag you would like to use to wrap the
+     * translations.
+     */
+    tagName?: string,
+};
+
+type State = {
+    composition: Composition,
+    id: string,
+    source: string,
+};
+
+/**
+ * Replace the text inside of this component with a translation. This
+ * component is built on top of react-intl, so it works along with the
+ * regular react-intl components and objects you are used to, and it gets
+ * its translations from react intl as well. The FormattedCompMessage component can
+ * be used wherever it is valid to put JSX text. In regular Javascript
+ * code, you should continue to use the intl.formatMessage() call and
+ * extract your strings into a message.js file.
+ */
+class FormattedCompMessage extends React.Component<Props, State> {
+    composition: Composition;
+
+    str: string;
+
+    static defaultProps = {
+        tagName: 'span',
+    };
+
+    constructor(props: Props) {
+        super(props);
+
+        // these parameters echo the ones in react-intl's FormattedMessage
+        // component, plus a few extra
+        const {
+            id, // the unique id of the string
+            defaultMessage, // The English string + HTML + components that you want translated
+            description, // An explanation of the message for the translators
+            count, // the pivot count to choose a plural form
+            children, // the components within the body
+        } = this.props;
+
+        const sourceElements = defaultMessage || children;
+
+        if (isDevEnvironment()) {
```

Comment:
Defining these props as required in the accompanying Flow types should be adequate to ensure they're passed in.